### ISSUE TITLE
a11y: widespread issues with font scaling

### ISSUE THREAD

**Issue by clshortfuse** [material-components/material-components-web#1932]
### What MDC-Web Version are you using?

master as of 65fe338fab3bae90ed89d3fb5e033c969562fd33

### What browser(s) is this bug affecting?

All (though testing on Chrome 63)

### What OS are you using?

Mac OS X

### What are the steps to reproduce the bug?

Use browser font accessibility options to increase font size or manually change a component's font-size.

### What is the expected behavior?

Components should properly allow for variable font sizes.

### What is the actual behavior?

Various rendering bugs such as: 

Element cropping:
![Element Cropping](https://i.imgur.com/roYNM6B.png)

Incorrect spacing:
![Incorrect spacing](https://i.imgur.com/Mjde4nX.png)

Oversized fonts (labels should be 12sp and padding should be consistent) :
![Oversided fonts](https://i.imgur.com/7IiQudl.png)

### Any other information you believe would be useful?

This deals with a much larger issue with how font scaling and component rendering is handled in general. I worked on the Angular Material (JS) team mostly focusing on reworking components to align spec visually and also accessibility concerns regarding large font sizes, contrast, and keyboard navigation.

You can see some of the work I've done on Angular Material in [this codepen](http://codepen.io/shortfuse/pen/qZojGW). I started with tackling the 12sp label issue and started working form there. Unfortunately, the some changes weren't merged due to the fact that project was sunsetted in favor of Angular Material 2.

I'm willing to fix the these same a11y issues here on MDC-Web, though I believe it's better have a discussion as to how to tackle the overall issue. Also, there's the possibility of using `rem` units instead of `px` for font rendering, which allow users/developers to scale the font sizes of all components.

**Comment by acdvorak** [comment]
Thanks for filing this issue and including such helpful screenshots! 🎉 

We definitely want to fix this, and it sounds like you probably know more about it than we do, so we would _really_ appreciate your help!

It seems like most of these issues should be fairly straightforward to fix. The only thing that _might_ be tricky is components that use CSS animations and/or have JavaScript that computes the size/position of elements (e.g., text-field floating labels), but I could certainly be wrong.

In terms of implementation, we might end up needing to create a set of Sass mixins and/or functions to convert between `dp`, `sp`, `rem`, `px`, etc. and call them whenever we set `margin`, `padding`, `line-height`, etc. in our components.

I'll loop in someone from the a11y team to see if they have any thoughts on best practices for units (`rem` vs. `px` vs. all the other ones) and how to scale padding/margins.

What's your advice on how to fix some of these issues? Would using different units (e.g., `rem`) solve a lot of them?

**Comment by clshortfuse** [comment]
There was a point in time where I lived and breathed the [Web Content Accessibility Guidelines 2.0 document](https://www.w3.org/TR/WCAG20/) and could easily explain it all off the top my head, but I'm a bit rusty now.

From what I remember, you'd first have to understand what the numbers meant in the Material Design guidelines.

* `px` would relate to Adobe Illustrator measurements which differ from CSS/SVG sometimes by a factor of `4.0` (ie: `shadowBlur` and `shadowOffset`). It's rarely to have to worry about this, but it's good to know if sometimes shadows don't match spec. Essentially `4px=1px`. 
* `dp`/`dip` relates density-independent-pixel which is commonly seen with Android. This is tied to the the browser's zoom level. Both content and text should scale with this value. We can directly map `1dp=1px` since browsers handle zoom on their own.
* `sp` relates to scale-independent-pixel also seen on Android. This relates to font size on the screen. If a user decides to set large fonts on their system (or in their browser setting), the font size should scale based on this. We can incorporate this in web in two ways: 
  * Using `pt`, the size of the font is based on the browser/OS ppi settings. The math is generally `16sp=12pt`. When a user changes the system-wide font scaling or browser font scaling the components will scale correctly. 
  * Using `rem`, we can size fonts based on the root element. The math is generally `16sp=1.0rem`. This allows developers to manually change the `font-size` in the `html` component and resize the text for all their objects.  

The real decision is based on using `rem` vs using `pt`. I've rarely seen websites use `pt`, so I would lean on `rem`, which has an added option for developers. (On that point, some of my client have requested per-user font scaling instead of just system-wide, so I've seen first-hand its practicality). Regardless, continuing to use `px` would have accessibility concerns.

So when you have a component and write it's font-size should be `16px` you're saying it should be 16 density-independent-pixels big. You actually want to say it's `1.0rem` which is 16 scale-independent-pixels. Things get a little hairier when you want to use something other than 16sp. For example 14sp is actually `0.875rem`. But since we're using Sass to compile our CSS, we can easily make a mixing to do the math for us:

```
@function sp($multiplier) {
  @return ($multiplier/16.0)+rem;
};
```

Which would allow use to create rules like:

```
button {
  font-size: sp(14);
}
```

You'd have to check every component one by one to see how they'd break when you incorporate these changes, but since components are isolated, it's easier to slowly incorporate these changes. But once you mentally hash out the complexities of `em`, `line-height`, and `px` and start incorporating them, you'll eventually find yourself fighting less with rendering engines and see things start to come together, eventually lining up with the reference images in the Material Design guideline site. You will likely also see a reduced need to use Javascript for alignment tweaking.

Of course, this is just my suggestion based on my prior efforts, but I'll concede to whatever the team decides is best.

PS: I've also joined the Discord group which is probably a better medium to discuss topic in-depth.

**Comment by clshortfuse** [comment]
I've made significant progress with fixing the the text fields. There's a lot of code I can clean up and I do have some idea of bringing support for floating labels and bottom-line to CSS-only.

https://codepen.io/shortfuse/pen/WdKBjg

I do see the .mdc-text-field element itself depends on an external `bottom-margin` and will work to make it more self-contained. Also, I've seen some uses of `opacity` instead of `rgba` values, which forces 3D rendering in some browsers.

I still have to apply some flex content for icons, but I don't foresee much of an issue.
